File: lib/kino/log.rb
# frozen_string_literal: true

module Kino
  # Server log lines: the lifecycle notices, crash and respawn reports,
  # hook failures, the failed-request report, and whatever apps write to
  # `rack.errors`, all in one shape:
  #
  #   kino[4213] worker-3: after_worker_boot hook raised RuntimeError: boom
  #
  # The label is syslog's `ident[pid]` tag plus the source that spoke: the
  # worker ractor and/or thread by name, `main` for neither. On color
  # terminals the label is dim, yellow, or red by level; the message stays
  # plain. Notes go to stdout, warnings and errors to stderr.
  #
  # Hooks may log through here too (`Kino::Log.info "cache warm"`). Every
  # method is safe inside a worker ractor: the line is handed to the
  # native layer, which owns the streams, so no ractor touches `$stdout`
  # or `$stderr` itself.
  module Log
    # Frames shown in a failed-request report before the rest are folded.
    FRAMES = 12

    # The working directory at boot, stripped from backtrace frames so the
    # app's own code reads `app/controllers/x.rb:9` rather than an
    # absolute path (frozen: worker ractors read it).
    WORKING_DIR = File.join(Dir.pwd, "").freeze

    module_function

    # @param message [#to_s]
    # @return [void]
    def info(message)
      Native.log_line("info", source, message.to_s)
    end

    # @param message [#to_s]
    # @return [void]
    def warn(message)
      Native.log_line("warn", source, message.to_s)
    end

    # @param message [#to_s]
    # @return [void]
    def error(message)
      Native.log_line("error", source, message.to_s)
    end

    # The failed-request report: the request line, the error, and where it
    # raised in the app, then the backtrace with the app's own frames
    # first (relative to the working directory) and the rest folded.
    #
    #   500 GET /boom · RuntimeError: kaboom (app.rb:12:in 'explode')
    #       app.rb:12:in 'explode'
    #       /gems/rack-3.2.7/lib/rack/builder.rb:...
    #       … 38 more
    #
    # @param error [Exception]
    # @param env [Hash] the Rack env of the failed request
    # @param status [Integer] the status the client got
    # @return [void]
    def exception(error, env, status: 500)
      frames, depth = trace(error)
      site = frames.first ? " (#{frames.first})" : ""
      lines = ["#{status} #{env["REQUEST_METHOD"]} #{env["PATH_INFO"]} · #{error.class}: #{error.message}#{site}"]
      frames.each { |frame| lines << "    #{frame}" }
      lines << "    … #{depth - frames.size} more" if depth > frames.size
      error(lines.join("\n"))
    end

    # The `kino[<pid>] <source>:` tag a line from here carries.
    # @return [String]
    def label
      "kino[#{Process.pid}] #{source}:"
    end

    # Who is speaking: the ractor's name, the thread's name, both joined
    # with a slash, or `main` when neither is named. Kino names its
    # worker ractors and threads `worker-N`.
    # @return [String]
    def source
      parts = [Ractor.current.name, Thread.current.name].compact
      parts.empty? ? "main" : parts.join("/")
    end

    # The backtrace as [frames, depth]: each frame relativized to the
    # working directory, the app's own frames floated to the front (the
    # raise site in your code reads first; gem and stdlib frames keep
    # their order below), capped at FRAMES; depth is the real length.
    def trace(error)
      raw = error.backtrace || []
      app, rest = raw.map { |frame| frame.delete_prefix(WORKING_DIR) }.partition { |frame| app_frame?(frame) }
      [(app + rest).first(FRAMES), raw.size]
    end

    # A project-relative path (the working-directory prefix came off, so
    # it does not start with `/`) that is not a synthetic frame (`(eval)`,
    # `<internal:...>`). Gem and stdlib frames stay absolute.
    def app_frame?(frame)
      !frame.start_with?("/", "<", "(")
    end

    private_class_method :trace, :app_frame?
  end
end

File: ext/kino/src/log.rs
//! Server log lines: lifecycle notices, crashes and respawns, hook
//! failures, the failed-request report, and whatever apps write to
//! rack.errors, all in one shape:
//!
//! ```text
//! kino[4213] worker-3: after_worker_boot hook raised RuntimeError: boom
//! ```
//!
//! The label is syslog's `ident[pid]` tag plus the source that spoke (the
//! ractor and/or thread name, `main` for neither), styled by level; the
//! message stays plain. Ruby builds the source, since only Ruby knows its
//! ractor and thread names, and hands the rest over; this side decides
//! color per stream and writes, so worker ractors never touch $stdout or
//! $stderr themselves. A multi-line message is labelled on its first line.

use std::io::Write;

use magnus::{Error, Ruby};

use crate::style::{self, Stream};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Level {
    Info,
    Warn,
    Error,
}

impl Level {
    /// The level named by Ruby ("info", "warn", "error").
    pub fn parse(name: &str) -> Option<Level> {
        match name {
            "info" => Some(Level::Info),
            "warn" => Some(Level::Warn),
            "error" => Some(Level::Error),
            _ => None,
        }
    }

    /// Notes go to stdout; warnings and errors to stderr.
    fn stream(self) -> Stream {
        match self {
            Level::Info => Stream::Stdout,
            Level::Warn | Level::Error => Stream::Stderr,
        }
    }

    fn sgr(self) -> &'static str {
        match self {
            Level::Info => style::DIM,
            Level::Warn => style::WARN,
            Level::Error => style::ERROR,
        }
    }
}

/// Write one line from `source` at `level`.
pub fn emit(level: Level, source: &str, message: &str) {
    let stream = level.stream();
    let label = label(std::process::id(), source);
    let line = format_line(level, &label, message, style::enabled(stream));
    match stream {
        Stream::Stdout => {
            let _ = writeln!(std::io::stdout().lock(), "{line}");
        }
        Stream::Stderr => {
            let _ = writeln!(std::io::stderr().lock(), "{line}");
        }
    }
}

/// The `kino[<pid>] <source>:` tag.
pub fn label(pid: u32, source: &str) -> String {
    format!("kino[{pid}] {source}:")
}

/// The label styled by level, then the message as given.
pub fn format_line(level: Level, label: &str, message: &str, color: bool) -> String {
    format!("{} {message}", style::sgr(level.sgr(), label, color))
}

/// The Ruby entry point (Kino::Log): Ruby knows its ractor and thread,
/// the native side knows the terminal.
pub fn log_line(ruby: &Ruby, level: String, source: String, message: String) -> Result<(), Error> {
    let level = Level::parse(&level).ok_or_else(|| {
        Error::new(
            ruby.exception_arg_error(),
            format!("unknown log level {level:?}"),
        )
    })?;
    emit(level, &source, &message);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::{format_line, label, Level};

    #[test]
    fn label_is_a_syslog_tag_plus_the_source() {
        assert_eq!(label(4213, "main"), "kino[4213] main:");
        assert_eq!(
            label(4213, "worker-3/thread-2"),
            "kino[4213] worker-3/thread-2:"
        );
    }

    #[test]
    fn plain_line_is_label_then_message() {
        assert_eq!(
            format_line(Level::Info, "kino[1] main:", "hello", false),
            "kino[1] main: hello"
        );
    }

    #[test]
    fn color_styles_only_the_label_by_level() {
        assert_eq!(
            format_line(Level::Info, "kino[1] main:", "hello", true),
            "\x1b[90mkino[1] main:\x1b[0m hello"
        );
        assert_eq!(
            format_line(Level::Warn, "kino[1] main:", "careful", true),
            "\x1b[33mkino[1] main:\x1b[0m careful"
        );
        assert_eq!(
            format_line(Level::Error, "kino[1] main:", "broke", true),
            "\x1b[91mkino[1] main:\x1b[0m broke"
        );
    }

    #[test]
    fn a_report_is_labelled_on_its_first_line_only() {
        assert_eq!(
            format_line(
                Level::Error,
                "kino[1] main:",
                "500 GET / · X: y\n    a.rb:1",
                false
            ),
            "kino[1] main: 500 GET / · X: y\n    a.rb:1"
        );
    }

    #[test]
    fn levels_parse_from_their_ruby_names() {
        assert_eq!(Level::parse("info"), Some(Level::Info));
        assert_eq!(Level::parse("warn"), Some(Level::Warn));
        assert_eq!(Level::parse("error"), Some(Level::Error));
        assert_eq!(Level::parse("debug"), None);
    }
}
